### ISSUE TITLE
Create a generic OpKernel which slides on input space with strides.

### ISSUE THREAD

**Issue by karandesai-96** [tiny-dnn/tiny-dnn#810]
While performing parameter refactor, I noticed that the following kernels can be merged into a generic one:

1. `conv2d`
2. `maxpool`
3. `avepool`
4. `global_avepool`
5. `deconv2d`

We can create a single kernel which knows the dimension of filter windows and its strides and it slides through the whole input space. It can accept `std::function` which will perform a certain operation with the elements within the window - for example dot_product for `conv2d`, max operation for `maxpool` etc.

Our `avepool` forward kernel works same as conv2d kernel currently (performing dot product) the difference being that in all test cases weights are initialized as 1 and bias as 0 which effectively reduces to average pooling.

If that is the case then we should reuse our code to ensure lesser maintenance. I am not sure upto what extent do they harm flexibility or extensibility. Open for discussion.

**Comment by karandesai-96** [comment]
Also there should be a generic OpKernel to pad and unpad Tensors. Currently there is a lot of code duplication within `conv2d` and `deconv2d`, while `avepool` layer asks for padding_type in input but I never saw it perform padding if `valid`.

I could not shift `avepool` to standalone `OpKernel` like others because they accept certain members of the worker storage. I can add a method in `backend.h` just like quantization kernels but that is not standard. We should think of passing worker specific storage to `OpKernelContext` as well. If this storage thing is not to be used in future then we should discuss about alternatives.

**Comment by edgarriba** [comment]
go ahead with this

**Comment by karandesai-96** [comment]
Sure, after the storm calms down. Me and @Randl are making a lot of changes with the layer and kernel definitions itself.

**Comment by edgarriba** [comment]
sounds good ! try to benchmark stuff to make sure that what change you do is to improve

**Comment by karandesai-96** [comment]
Right, I will write benchmarks for all kernels first then start making changes..
